src/bam/simulate_reads.rs

```diff
@@ -0,0 +1,165 @@
+    artificial_record.set_flags(record.flags());
+    artificial_record.set_insert_size(record.insert_size());
+    artificial_record.set_mapq(record.mapq());
+    Ok(artificial_record)
+}
+fn build_sequence(
+    reference: &[u8],
+    artificial_reference: &[u8],
+    record: &bam::Record,
+    offset: usize,
+) -> Result<Vec<u8>> {
+    let mut artificial_seq = Vec::new();
+    let record_seq = record.seq().as_bytes();
+    let mut record_pos = 0;
+    let mut ref_pos = record.pos() as usize - offset;
+    //Create random seq for leading softclips
+    for cigar in record.cigar_cached().unwrap().iter() {
+        match cigar.char() {
+            'S' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq)?;
+                record_pos += cigar.len() as usize;
+            }
+            'M' => {
+                (0..cigar.len()).for_each(|_| {
+                    let ref_base = artificial_reference.get(ref_pos).unwrap();
+                    if record_seq.get(record_pos).unwrap() == reference.get(ref_pos).unwrap() {
+                        artificial_seq.push(*ref_base);
+                    } else {
+                        add_random_base(&mut artificial_seq, ref_base);
+                    }
+                    ref_pos += 1;
+                    record_pos += 1;
+                });
+                // Add reference bases except for mismatches
+            }
+            'I' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq)?;
+                record_pos += cigar.len() as usize;
+            }
+            'D' | 'N' => {
+                ref_pos += cigar.len() as usize;
+            }
+            'X' => {
+                (0..cigar.len()).for_each(|_| {
+                    let ref_base = artificial_reference.get(ref_pos).unwrap();
+                    add_random_base(&mut artificial_seq, ref_base);
+                    ref_pos += 1;
+                });
+                record_pos += cigar.len() as usize;
+            }
+            '=' => {
+                (0..cigar.len()).for_each(|_| {
+                    let ref_base = artificial_reference.get(ref_pos).unwrap();
+                    artificial_seq.push(*ref_base);
+                    ref_pos += 1;
+                });
+                record_pos += cigar.len() as usize;
+            }
+            _ => {}
+        }
+    }
+
+    Ok(artificial_seq)
+}
+
+fn add_random_base(seq: &mut Vec<u8>, excluded_base: &u8) {
+    let mut rng = rand::thread_rng();
+    let mut alphabet = vec![b'A', b'C', b'G', b'T'];
+    alphabet.retain(|x| x != excluded_base);
+    let idx = rng.gen_range(0, 3);
+    seq.push(alphabet[idx]);
+}
+
+fn add_random_bases(length: u64, seq: &mut Vec<u8>) -> Result<()> {
+    let alphabet = [b'A', b'C', b'G', b'T'];
+    let mut rng = rand::thread_rng();
+    (0..length).for_each(|_| {
+        let idx = rng.gen_range(0, 4);
+        seq.push(alphabet[idx])
+    });
+    Ok(())
+}
```

Comment:
I would unify these two to avoid the duplication. Also, I would provide them with an RNG that is initialized before instead of initializting a new one each time.

Reply:
This should be done now.